File: ex3/python/variable_blinking_period.py
#!/usr/bin/python3
#**
#******************************************************************************
#* @file        /ex3/python/variable_blinking_period.py
#* @author      Radoslaw Goralewski
#* @version     V1.0
#* @date        26-Mar-2020
#* @brief       Raspberry Pi, Serwer IoT Lab : Exercise 3, Python script GPIO
#******************************************************************************

import sys
import getopt
from parse import parse
import time
import select
try:
    import RPi.GPIO as GPIO
except RuntimeError:
    print("Error importing RPi.GPIO! This is probably because you need superuser privileges. You can achieve this by using 'sudo' to run your script.")

# Default values
min_period = 0.1
max_period = 0.5
step = 0.1

# Parse input
sysarg = sys.argv[1:]
try:
    opts, args = getopt.getopt(sysarg, "m:M:s:")
except getopt.GetoptError as e:
    print(e)
    sys.exit(1)

# Check options
for opt, arg in opts:
    if opt in '-m':
        min_period = float(arg)
    elif opt in '-M':
        max_period = float(arg)
        if (max_period <= min_period):
            print("Maximum period must be greater than minimum period!")
    elif opt in '-s':
        step = float(arg)

# Print set values
print("Minimum period: {}, Maximum period: {}, Step: {}".format(min_period, max_period, step))

# Initialize necessary variables
blinking_period = max_period
buttonState = True
buttonStateLast = True
ledState = False
i = ''

# Pin Definitions:
ledPin = 16 # < LED: Physical pin 16, BCM GPIO23
buttonPin = 18 # < Push-button: Physical pin 18, BCM GPIO24

# Pin Setup:
GPIO.setmode(GPIO.BOARD)
GPIO.setup(ledPin, GPIO.OUT) # LED pin set as output
GPIO.setup(buttonPin, GPIO.IN) # Button pin set as input

GPIO.output(ledPin, ledState)

print("Press ENTER to exit.")

while not i:
    # Waiting for I/O completion
    i, o, e = select.select([sys.stdin], [], [], blinking_period)

    if i:
        sys.stdin.readline()
        GPIO.cleanup() # cleanup all GPIO
        exit()

    # If button is pressed, change blinking period
    buttonState = (GPIO.input(buttonPin) == GPIO.HIGH)
    if (buttonState is False) and (buttonStateLast is True):
        blinking_period -= step
        if blinking_period < min_period:
            blinking_period = max_period
        print("Button pressed! Current period: {:.1f}".format(blinking_period))
    # Save last button state
    buttonStateLast = buttonState

    # Change LED state
    ledState = not ledState
    GPIO.output(ledPin, ledState)

File: ex3/cpp/variable_blinking_period.cpp
/**
 ******************************************************************************
 * @file       /ex3/cpp/variable_blinking_period.cpp
 * @author     Radoslaw Goralewski
 * @version    V1.0
 * @date       28-Mar-2020
 * @brief      Raspberry Pi, Serwer IoT Lab : Exercise 3, C++ program GPIO
 ******************************************************************************
 */

#include <iostream>
#include <unistd.h>
#include <stdio.h>
#include <wiringPi.h>
#include <future>
#include <thread>
#include <chrono>

int main(int argc, char *argv[])
{
    //Default values
    int min_period = 100;
    int max_period = 500;
    int step = 100;


    int arg;

    /* Standard while / switch procedure for ' getopt ' function */
    while ((arg = getopt(argc, argv, "m:M:s:")) != -1)
    {
        switch(arg)
        {
            case 'm':
                min_period = atof(optarg) * 1000;
                break;
            case 'M':
                max_period = atof(optarg) * 1000;
                break;
            case 's':
                step = atof(optarg) * 1000;
                break;
            case '?':
                if((optopt == 'm') || (optopt == 'M') || (optopt == 's'))
                    std::cerr << "option -" << static_cast<char>(optopt) << " requires argument" << std::endl;
                else if(isprint(optopt))
                    std::cerr << "option -" << static_cast<char>(optopt) << " not recognized" << std::endl;
                else
                    std::cerr << "option character \\x" << optopt << " not recognized" << std::endl;
                return 1;
            default:
                abort();
        }
    }

    std::cout << "Minimum period: " << float(min_period / 1000.0) << ", Maximum period: " << float(max_period / 1000.0) << ", Step: " << float(step / 1000.0) << std::endl;

    //Initialize necessary variables
    bool button_state = true;
    bool button_state_last = true;
    bool led_state = false;

    const int led = 4; // < LED: Physical pin 16, BCM GPIO23, and WiringPi pin 4
    const int button = 5; // < LED: Physical pin 18, BCM GPIO24, and WiringPi pin 5

    std::chrono::milliseconds blinking_period(max_period);
    std::future<int> async_getchar = std::async(std::getchar);

    wiringPiSetup();

    pinMode(led, OUTPUT);
    pinMode(button, INPUT);

    digitalWrite(led, LOW);

    std::cout << "Press ENTER to exit." << std::endl;

    while(1)
    {
        //Read button state
        button_state = (digitalRead(button) == HIGH);

        //If was pressed, change the blinking period
        if (!button_state && button_state_last)
        {
            blinking_period -= std::chrono::milliseconds(step);
            if (blinking_period < std::chrono::milliseconds(min_period)) {
                blinking_period = std::chrono::milliseconds(max_period);
            }
            std::cout << "Button pressed! Current period: " << float(blinking_period.count() / 1000.0) << std::endl;
        }

        //Save state
        button_state_last = button_state;

        //Toggle LED
        led_state = !led_state;
        digitalWrite(led, led_state);

        //Wait for the blinking_period time, if enter was pressed exit program
        if(async_getchar.wait_for(blinking_period) == std::future_status::ready)
        {
            async_getchar.get();
            break;
        }
    }

    return 0;
}
